src/suggestion-box.ts

```diff
@@ -0,0 +1,409 @@
+                >
+                    <div style="display:${this.show_empty ? "block" : "none"}">
+                        <slot id="sgn-empty" name="sgn-empty"></slot>
+                    </div>
+                    <div
+                        style="display:${!this.show_empty ? "block" : "none"};"
+                        @click=${this._saveInputEvent}
+                        @mouseover=${this._mouseOverEvent}
+                    >
+                        <slot id="private-item" name="__private-item"></slot>
+                    </div>
+                </div>
+            </div>
+        `;
+    }
+    private _indexOfTarget(event: any) {
+        let el = event.target;
+        while (el) {
+            if (el.slot == "__private-item") break;
+            el = el.parentElement;
+        }
+        return this.private_items.indexOf(el);
+    }
+    private _saveInputEvent(event: Event) {
+        let indextarget = this._indexOfTarget(event);
+        if (this.result[indextarget]) {
+            (this.input_el as HTMLInputElement).value =
+                this.result[indextarget][this.listkey];
+            this.show_container = false;
+        } else if (indextarget === this.result.length) {
+            this._loadmoreEntry(event);
+        }
+    }
+    private _mouseOverEvent(e: any) {
+        let curr_index = 0;
+        curr_index = this._indexOfTarget(e);
+        this.index = curr_index;
+        if (this.index >= 0) {
+            const prevItem = this.private_items[this.index];
+            if (prevItem) {
+                prevItem.classList.remove("sgn-active");
+            }
+        }
+        if (curr_index >= 0) {
+            this.private_items[this.index].classList.add("sgn-active");
+        }
+        this.prevIndex = curr_index;
+    }
+    private _inputSelectEvent() {
+        this.index = -1;
+    }
+    private _inputFocusEvent() {
+        if (this.show_empty) this.show_container = true;
+        this.index = -1;
+    }
+    private _mouseOutEvent() {
+        if (this.prevIndex >= 0 && this.prevIndex < this.result.length) {
+            this.private_items[this.index]?.classList.remove("sgn-active");
+        }
+    }
+    windowClickEvent(event: any) {
+        let el = event.target;
+        while (el) {
+            if (el == this) return;
+            el = el.parentElement;
+        }
+        this.show_container = false;
+    }
+    private _format(str: any, obj: any, keys: string) {
+        const input = (this.input_el as HTMLInputElement)?.value;
+        if (typeof obj == "string" || typeof obj == "number") return obj;
+        else {
+            for (let key of Object.keys(obj)) {
+                let template: any;
+                if (key === keys) {
+                    template =
+                        this.highlight === "matched"
+                            ? this._highlight(obj[key], input.toLowerCase())
+                            : this._startsWithHighlight(obj[key]);
```

Comment:
What if highlight neither matched nor end. What if its null, then we are default highlighting the end part alone.